cpp/include/raft/core/operators.hpp

```diff
@@ -0,0 +1,292 @@
+  constexpr RAFT_INLINE_FUNCTION Type operator()(const Type& a, const Type& b) const
+  {
+    if (b > a) { return b; }
+    return a;
+  }
+};
+
+struct sqdiff_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(const Type& a, const Type& b) const
+  {
+    Type diff = a - b;
+    return diff * diff;
+  }
+};
+
+struct argmin_op {
+  template <typename KVP>
+  constexpr RAFT_INLINE_FUNCTION KVP operator()(const KVP& a, const KVP& b) const
+  {
+    if ((b.value < a.value) || ((a.value == b.value) && (b.key < a.key))) { return b; }
+    return a;
+  }
+};
+
+struct argmax_op {
+  template <typename KVP>
+  constexpr RAFT_INLINE_FUNCTION KVP operator()(const KVP& a, const KVP& b) const
+  {
+    if ((b.value > a.value) || ((a.value == b.value) && (b.key < a.key))) { return b; }
+    return a;
+  }
+};
+
+template <typename ScalarT>
+struct const_op {
+  const ScalarT scalar;
+
+  constexpr const_op(const ScalarT& s) : scalar{s} {}
+
+  template <typename... Args>
+  constexpr RAFT_INLINE_FUNCTION ScalarT operator()(Args...) const
+  {
+    return scalar;
+  }
+};
+
+template <typename ScalarT, typename BinaryOpT>
+struct scalar_op {
+  const ScalarT scalar;
+  const BinaryOpT composed_op;
+
+  template <typename OpT     = BinaryOpT,
+            typename UnusedT = std::enable_if_t<std::is_default_constructible_v<OpT>>>
+  constexpr scalar_op(const ScalarT& s) : scalar{s}, composed_op{}
+  {
+  }
+  constexpr scalar_op(const ScalarT& s, BinaryOpT o) : scalar{s}, composed_op{o} {}
+
+  template <typename InT>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(InT a) const
+  {
+    return composed_op(a, scalar);
+  }
+};
+
+template <typename Type>
+using scalar_add_op = scalar_op<Type, add_op>;
+
+template <typename Type>
+using scalar_sub_op = scalar_op<Type, sub_op>;
+
+template <typename Type>
+using scalar_mul_op = scalar_op<Type, mul_op>;
+
+template <typename Type>
+using scalar_div_op = scalar_op<Type, div_op>;
+
+template <typename Type>
+using scalar_div_checkzero_op = scalar_op<Type, div_checkzero_op>;
+
+template <typename Type>
+using scalar_pow_op = scalar_op<Type, pow_op>;
+
+template <typename OuterOpT, typename InnerOpT>
+struct compose_op {
+  const OuterOpT outer_op;
+  const InnerOpT inner_op;
+
+  template <typename OOpT    = OuterOpT,
+            typename IOpT    = InnerOpT,
+            typename UnusedT = std::enable_if_t<std::is_default_constructible_v<OOpT> &&
+                                                std::is_default_constructible_v<IOpT>>>
+  constexpr compose_op() : outer_op{}, inner_op{}
+  {
+  }
+  constexpr compose_op(OuterOpT out_op, InnerOpT in_op) : outer_op{out_op}, inner_op{in_op} {}
+
+  template <typename... Args>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(Args&&... args) const
+  {
+    return outer_op(inner_op(std::forward<Args>(args)...));
+  }
+};
+
+template <typename OuterOpT, typename InnerOpT1, typename InnerOpT2 = raft::identity_op>
```

Comment:
As you suggested, we could generalize this one as well (NB: add include the tuple header):
```C++
template <typename OuterOpT, typename... ArgOpsT>
struct map_args_op {
  const OuterOpT outer_op;
  const std::tuple<ArgOpsT...> arg_ops;

  constexpr explicit map_args_op(OuterOpT outer_op, ArgOpsT... arg_ops)
    : outer_op{outer_op}, arg_ops{arg_ops...}
  {
  }

  template <typename... Args>
  constexpr RAFT_INLINE_FUNCTION auto operator()(Args&&... args) const
  {
    constexpr size_t kNumOps = sizeof...(ArgOpsT);
    static_assert(kNumOps == sizeof...(Args),
                  "The number of arguments does not match the number of mapping operators");
    return map_args(std::make_index_sequence<kNumOps>{}, std::forward<Args>(args)...);
  }

 private:
  template <size_t... I, typename... Args>
  constexpr RAFT_INLINE_FUNCTION auto map_args(std::index_sequence<I...>, Args&&... args) const
  {
    return outer_op(std::get<I>(arg_ops)(std::forward<Args>(args))...);
  }
};
```